### ISSUE TITLE
Gremlin HTTP Api: HashMap should not cast to Element

### ISSUE THREAD

**Issue by marcelalburg** [orientechnologies/orientdb#8382]
## Expected behavior  
When i use the HTTP Api with the gremlin command:

`g.V().has("key", "<key>").repeat(__.in("is_part_of").simplePath()).emit().path().by(label()).groupCount().values()
`
it should return a Json HashMap.

## Actual behavior  
Exception:

`ava.lang.ClassCastException: java.util.HashMap cannot be cast to org.apache.tinkerpop.gremlin.structure.Element`

**Comment by maggiolo00** [comment]
Hi @marcelalburg 

do you have the entire stacktrace of the exception?

Thanks

**Comment by marcelalburg** [comment]
Sorry, no. Because i use the HTTP Api and even in the Server Logs. there is no stracktrace

```
curl -u <user>:<pwd> -s -XPOST http://localhost:2480/command/belayus/gremlin -d '{"command": "g.V().has(\"key\", \"<key>\").repeat(__.in(\"is_part_of\").simplePath()).emit().path().by(label()).groupCount()"}' | jq '.'
{
  "errors": [
    {
      "reason": 500,
      "code": 500,
      "content": "java.lang.ClassCastException: org.apache.tinkerpop.gremlin.process.traversal.step.util.MutablePath cannot be cast to java.lang.String"
    }
  ]
}
```

if i add the "next()" inside this HTTP command, i got an empty result.

The Java Api is working as expected:

```
Map<Object, Long> data = g.V().has(primaryKeyName, primaryKey.toString())
   .repeat(__.in("is_part_of").simplePath()).emit()
  .path().by(__.label()).groupCount().next();
```

**Comment by maggiolo00** [comment]
@marcelalburg 

do you have a simple dataset where i can reproduce this and debug it?

Thanks

**Comment by marcelalburg** [comment]
I wrote a test:

https://github.com/marcelalburg/orientdb3-embedded/blob/master/src/test/java/example/GremlinVsHttpTest.java#L308

The original query is inside and the query what should work (the java api is working).

Sorry, that i wrote so much tickets. Currently i work on a Prototype to move a big application from Neo4J to OrientDb and i was expecting that i could copy the gremlin commands easily to orientdb but there are much differences ( maybe because of tinkerpop version 3).

Thanks for your help.

**Comment by maggiolo00** [comment]
@marcelalburg 

was the application with Tinkerpop2 and Java APis?

**Comment by marcelalburg** [comment]
I think so, i just have the code and raw data of the HTTP calls and not the entiere application.

**Comment by marcelalburg** [comment]
The same with this curl:

```
curl -u <user>:<password> -s -XPOST http://localhost:2480/command/<db>/gremlin -d '{"command": "g.V().has(\"key\", \"key\").repeat(__.in(\"is_part_of\").simplePath()).emit().path().by(label()).groupCount().unfold()"}' | jq '.'
{
  "errors": [
    {
      "reason": 501,
      "code": 501,
      "content": "java.lang.UnsupportedOperationException: Cannot convert [x, y]=19 - class java.util.HashMap$Node to JSON"
    }
  ]
}
```

**Comment by maggiolo00** [comment]
Hi @marcelalburg 

i've reproduced this. Seems it's not an OrientDB problem

I've tried this in a GremlinConsole

```
graph = TinkerFactory.createModern()
root = graph.addVertex(T.label, "example", "name", "root", "key", 1);
link = graph.addVertex(T.label, "example", "name", "other_node", "key", 2);
link2 = graph.addVertex(T.label, "example2", "name2", "other_node2", "key", 3);
link2.addEdge("contains", link);
link.addEdge("contains", root);
g = graph.traversal()
g.V().has('key', 1).repeat(__.in("contains").simplePath()).emit().path().by(label()).groupCount().values()
```

And still got the Exception. I'm not sure `values()` works well with `groupCount()`

**Comment by marcelalburg** [comment]
The exception happen only with the HTTP client. Not if you use the Thinkerpop implementation.

You can see this in the response of this test: https://github.com/marcelalburg/orientdb3-embedded/blob/master/src/test/java/example/GremlinVsHttpTest.java#L308

**Comment by maggiolo00** [comment]
@marcelalburg 

what do you mean Tinkerpop implementation.
 Which is the example that does not fail?

**Comment by marcelalburg** [comment]
Hello,

i wrote a test "testGroupCountNativeGremlin" which uses Thinkerpop API (i called it native api)
https://github.com/marcelalburg/orientdb3-embedded/blob/master/src/test/java/example/GremlinVsHttpTest.java#L314

in this test, i need to call "next()" to get all data to my localVariable. This test works fine for me and the expected data comes back.

---

in the test "testGroupCount" i use exact the same query but i send it to the orientDB gremlin HTTP api:

https://github.com/marcelalburg/orientdb3-embedded/blob/master/src/test/java/example/GremlinVsHttpTest.java#L332

 i try to call .toList() and .next() but this returns me an empty list. 

.values() returns me the exception:

```
{
  "errors": [
    {
      "reason": 500,
      "code": 500,
      "content": "java.lang.ClassCastException: java.util.HashMap cannot be cast to org.apache.tinkerpop.gremlin.structure.Element"
    }
  ]
}
```

the question is now, how can i get the same result from the HTTP APi as from the TinkerPop native API ?

**Comment by maggiolo00** [comment]
@marcelalburg 

so in native you did not use `values()`. 

if you use it in native

```
Object data = g.V().has("key", ROOT_KEY.toString()).repeat(__.in("contains").simplePath()).emit().path().by(__.label())
					.groupCount().values().next();
```

You will get the same exception

**Comment by marcelalburg** [comment]
Yes, you're right. And how can i get the correct/same Response from the HTTP call ?

**Comment by maggiolo00** [comment]
let me check

**Comment by maggiolo00** [comment]
Hi @marcelalburg 

i've found the problem with `.next()`. I'm pushing a fix. 
Just a question. Do you need to use ODB Http API + Gremlin?

**Comment by marcelalburg** [comment]
Hello, thanks for finding the issue.

Yes, i need to use the HTTP Api with gremlin Commands. This has multiple reasons.

- my customer want's to have this :)
- the customer likes to be flexible as possible and independent from database behind (this is why i need to use thinkerpop 3 / gremlin)
  - the service should be as small as possible and should have less 3rd party dependencies (one reason why i need to use http)
  - maybe, my Java prototype will just used as prototype and after this the customer likes to switch to nodejs (copy paste the http requests and this should be fine)
- we'll deploy in an amazon/google cloud and it's easier to scale and more flexible, when the LoadBalancer routes the traffic between the nodes 
  -  the customer has not so much trust in the hazelcast - cloud deployment and thinks it's better to use the loadbalanced solution

This are the main reasons, why i need to use this setup. I hope this answers are helpful for you

Marcel

**Comment by maggiolo00** [comment]
Hi @marcelalburg 

that what i was thinking. The problem is that by using  using OrientDB gremlin endpoint, you will not be database independent. The HTTP JSON result is derived from OrientDB result set. That is different from  the Apache Tinkerpop Standard. http://tinkerpop.apache.org/docs/3.3.3/dev/io/

if you want to be standard over the protocol have you checked the Gremlin Server?

Otherwise if you still want to go with with the custom OrientDB endpoint we can expose a new one compatible with the IO reference of the Gremlin Server.

**Comment by marcelalburg** [comment]
Wow, thanks for that info. i'll check this project. But i wrote already a OrientDb Parser which fits my need for the prototype. For other Databases i just need to implement a new parser. But you info is very helpful.

Thx

**Comment by marcelalburg** [comment]
Hello,

i run my test and i see a result with "next" now. Can you confirm, that this issue is fixed with 3.0.4 ?

M

**Comment by maggiolo00** [comment]
Hi @marcelalburg 

it should be available in 3.0.4

**Comment by marcelalburg** [comment]
Ok, my test works as expected. Thanks a lot.
